### DISCUSSION TITLE
Warning Banner/Card in Canopy

### DISCUSSION THREAD

**Discussion by owensgit** [Legal-and-General/canopy#289] [Ideas] [answered]
I'm wondering how we can achieve the following designs in Canopy. This is a warning message/banner, but it is somewhat different to the current Alert component. Note that in the designs, the colour treatment is wrong - that aspect should be the same as the alert component. I can think of a few approaches:

### Approaches

**New Component** 
A brand new component, with content projection slots for the icon, message and button, using the same variant styles as the alert/detail etc.

**Updates to existing Alert Component**
We could add inputs to apply rounded corners/drop shadow,, use padding directives along with another things to get the result, content projection slot for the button on the right - or something along those lines.

**Updates to existing Card Component**
Add ability to change the background colour on cards, and maybe new content projection slots for the placement of the button and icon - or something along those lines

My current preference is to go with a new component, as it means less complexity in existing ones, but happy to discuss 👍 

### Design Questions

- Will this be used elsewhere?
- ~The background colour is a darker than the warning variant of the alert/message component - is this intentional?~
- Will there be other variants/colours of this design?
- How else might this component be used?

### Screenshots

Note the background colour is incorrect in these designs, it should be `--color-dandelion-yellow-lightest`, same as the warning variant.

**Desktop**
<img width="949" alt="Screenshot 2021-03-29 at 20 07 59" src="https://user-images.githubusercontent.com/1943532/112887329-f39edb80-90ca-11eb-88b5-1ac8644b9328.png">

**Tablet**
<img width="738" alt="Screenshot 2021-03-29 at 20 08 20" src="https://user-images.githubusercontent.com/1943532/112887335-f4d00880-90ca-11eb-92bc-a0848a1e8668.png">

**Mobile**
<img width="329" alt="Screenshot 2021-03-29 at 20 08 38" src="https://user-images.githubusercontent.com/1943532/112887339-f5689f00-90ca-11eb-92db-277ec69d81a3.png">

**Comment by alexcanning** [comment]
**Will this be used elsewhere?**
Not sure yet - potentially on product pages too?

**The background colour is a darker than the warning variant of the alert/message component - is this intentional?**
No, it should be a lighter yellow - `--color-dandelion-yellow-lightest` 

**Will there be other variants/colours of this design?**
None that have been documented yet, as far as I know.

**How else might this component be used?**
Can't see it being used for anything other than presenting a warning inline with other cards.

It doesn't look like an ~alert~ inline message component on it's own - it looks like a card with some warning styling on it. It's also worth noting that button variation doesn't exist yet. The mobile left margin looks like it needs some adjustment too.

**Reply by owensgit** [reply]
> it looks like a card with some warning styling on it

Can you see this becoming a thing in Canopy? E.g. having a card with a solid background colour, and then coloured text - a bit like the alert/inline message component?

> that button variation doesn't exist yet

It looks like an [outline primary button](https://legal-and-general-canopy.netlify.app/?path=/story/components-button--outline-primary), in a different colour. Do you see there being more types of this button for different background colours? Or this is a one off?

**Reply by alexcanning** [reply]
> Can you see this becoming a thing in Canopy? E.g. having a card with a solid background colour, and then coloured text - a bit like the alert/inline message component?

Possibly - seeing some other concepts where folks are just using other colours from the library (not necessarily inline message colours), but these haven't necessarily been approved yet.

> It looks like an outline primary button, in a different colour. Do you see there being more types of this button for different background colours? Or this is a one off?

Tinted outline buttons may need to be a thing if we're dropping them onto colour backgrounds, yeah.

**Reply by michelleallendavey** [reply]
Bit of background on that mobile spacing @alexcanning – when it's in the context of the PSP cards, Lisa was keen that it matched up to the Quick Action spacing above it, if that makes sense, as it was similar content (small icon, text to right) but out of kilter:

<img width="370" alt="Screenshot 2021-04-06 at 14 37 34" src="https://user-images.githubusercontent.com/70210216/113719912-05453c00-96e6-11eb-941f-a836bf5f74e4.png">

**Comment by elenagarrone** [comment]
I agree with you @owensgit, I think it should be a different component as there are quite a few differences with both the card and the alert component.

If we use either the existing alert or the card component we would have to add a new input which I think starts to make it overcomplicated.

**Comment by paul-request** [comment]
Does it need to be a new component at all? could you not just compose it using an alert wrapped in a card?

```html
<lg-card>
  <lg-card-content>
     <lg-alert />
  </lg-card-content>
</lg-card>
```

**Reply by paul-request** [reply]
sounds like this may not get re-used anywhere else, so perhaps not worth having a dedicated component?

**Reply by aaronBery** [reply]
I like this idea of creating a composition of components. As long as the building block components are only being extended to accommodate potential changes this is preferable to introducing more components.

**Comment by andy-polhill** [comment, marked as answer]
With regards to the button placement I wondered wether you could build this already by putting a nested grid inside a card component? 

@alexcanning  and I were discussing Components vs Patterns a bit the other day where we defined them as something like...

**Component** A low level component with a single area of concern.
**Pattern** A combination of components to recreate a common approach. 

I worry about creating a proliferation of very similar components, both at the design and implementation phase. With this I wonder if it's more of a pattern than a component.

The colour control could then be done by a directive potentially? or a card input property.

**Reply by andy-polhill** [reply]
```html
<lg-card variant="warning">
  <div lgContainer>
     <div lgRow>
       <div [lgCol]="12" [lgColMd]="6">
         Text
       </div>
       <div [lgCol]="12" [lgColMd]="6">
         <button>
       </div>
    </div>
  </div>
</div>
```

**Reply by elenagarrone** [reply]
We could have a directive that changes any element to have the `border-radius` and `drop-shadow` and then we would just use the alert component for the rest.

I agree on not building a lot of similar components but I also don't want to bloat existing ones with lots of different tiny variations that require more inputs 🤔 

Using the structure like @paul-request might work but would need the card padding to be modified. Unless we create a new card inner component e.g. `lg-card-alert`?

**Reply by andy-polhill** [reply]
we could use the padding directives to suppress the padding 🤔
